src/core/bpf-firewall.c

```diff
@@ -46,10 +49,439 @@ enum {
 };
 
 enum {
-        ACCESS_ALLOWED = 1,
-        ACCESS_DENIED  = 2,
+        ADDRESS_ACCESS_ALLOWED = (1 << 0),
+        ADDRESS_ACCESS_DENIED  = (1 << 1),
+        PORT_ACCESS_ALLOWED    = (1 << 2),
+        PORT_ACCESS_DENIED     = (1 << 3),
 };
 
+static void fix_jumps(struct bpf_insn *prog, uint length, uint offset) {
+        uint i;
+
+        /* Make all -1-length jumps go to the end of the block. */
+        for (i = 0; i < length; i++)
+                if ((prog[i].code & BPF_JMP) && (prog[i].off == -1))
+                        prog[i].off = offset - i - 1;
+}
+
+static int add_port_lookup_instructions(
+                BPFProgram *p,
+                int map_fd,
+                int protocol,
+                bool is_ingress,
+                int verdict) {
+        /*
+         * We are interested only in destination port regardless of ingress or
+         * egress -- source port is typically ephemeral. TCP and UDP have the
+         * ports in the same place in header structure.
+         */
+        int tcp_port_offset = offsetof(struct tcphdr, dest);
+
+        struct bpf_insn port_check_insn[] = {
+                /*
+                 * Assumptions: R2 contains the IP header length. R9 has the
+                 * transport layer protocol. R6 has the SKB.
+                 */
+
+                /*
+                 * The packet might have either an unknown transport layer
+                 * protocol or otherwise the extension headers can be broken
+                 * to the extent that we couldn't process them all in our
+                 * state machine. This means that we have really no choice but
+                 * to take the conservative estimate. We therefore split the
+                 * packets into three bins:
+                 *
+                 *  1. The protocols that we are protecting. Currently the known
+                 *     protocols are TCP and UDP. We'll look at the ports and
+                 *     accept/deny the packets based on the configuration rules.
+                 *  2. The protocols that we know but won't protect. Examples
+                 *     are ICMP and ICMPv6. Those we just pass through.
+                 *  3. The protocols that we don't know about. It's possible
+                 *     that the packet is an attempt to pass the firewall,
+                 *     because kernel might have different processing that we
+                 *     do. We drop those packets.
```

Comment:
This is bit problematic issue, because if our parsing of the packet somehow fails, the "non-recognized" protocol is then passed through. Even if I think that all the bases are covered (for example, the IPv6 extension headers are parsed correctly), there might be corner cases in which the kernel packet processing goes differently than ours. That might create a security issue.

For example, consider what happens if (hypothetically) a new IPv6 extension header gets introduced and supported in the kernel. Our parser will think that the packet is not TCP or UDP (because we only see the unknown extension header protocol), but the kernel might eventually pass it to the TCP or UDP socket anyway.

Maybe this should be configurable somehow. For instance, we could have a protocol option "other-protocols" or similar:

    IPPortIngressDeny="any/tcp any/udp other-protocols"
    IPPortIngressAllow="80/tcp"